File: Shortest_word_distance_II.java
class Solution{
  HashMap<String, List<Integer>> map;
    
  public WordDistance(String[] wordsDict){
      map = new HashMap<>();
      for(int i = 0; i < wordsDict; i++){
          if(!map.containsKey(wordsDict[i])){
              map.put(wordsDict[i], new ArrayList<>());
          }
          map.get(wordsDict([i]).add(i);
       }
  }
                  
  public int shortest(String word1, String word2){
      List<Integer> li1 = map.get(word1);
      List<Integer> li2 = map.get(word2);
      
      int i1 = 0;
      int i2 = 0;
      
      int min = Integer.MAX_VALUE;
      while(i1 < li1.size() && i2 < li2.size()){
          int index1 = li1.get(i1);
          int index2 = li2.get(i2);
          
          min = Math.min(min, Math.abs(index2 - index1));
          if(index1 < index2){
              i1++;
          }else{
              i2++;
          }
      }
      return min;
    }
  }
                  
//TC:O(m+k)
//SC: O(n)
                
        
                  
        

File: shortest_word_distance_ii.py
class WordDistance:
    from collections import defaultdict
    ## T.C = O(m+n)
    ## S.C = O(n)
    
    def __init__(self, wordsDict: List[str]):
        self.wordsDict = wordsDict
        self.hm = defaultdict(list)
        self.build_hm()

    def build_hm(self):
        for i in range(len(self.wordsDict)):
            self.hm[self.wordsDict[i]].append(i)

    def shortest(self, word1: str, word2: str) -> int:
        p1, p2 = 0, 0
        mn = float('inf')
        indices1 = self.hm[word1]
        indices2 = self.hm[word2]

        while p1 < len(indices1) and p2 < len(indices2):
            mn = min(mn, abs(indices1[p1] - indices2[p2]))
            if indices1[p1] < indices2[p2]:
                p1 += 1
            else:
                p2 += 1

        return mn



# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(wordsDict)
# param_1 = obj.shortest(word1,word2)